cmd/sovereignnode/dataCodec/dataCodec.go

```diff
@@ -0,0 +1,491 @@
+		marshaller: args.Marshaller,
+	}, nil
+}
+
+// SerializeEventData will receive an event data and serialize it
+func (dc *dataCodec) SerializeEventData(eventData sovereign.EventData) ([]byte, error) {
+	var gasLimit any
+	var function any
+	var args any
+	if eventData.TransferData != nil {
+		gasLimit = abi.U64Value{Value: eventData.GasLimit}
+		function = abi.BytesValue{Value: eventData.Function}
+		arguments := make([]any, len(eventData.Args))
+		for i, arg := range eventData.Args {
+			arguments[i] = abi.BytesValue{Value: arg}
+		}
+		args = abi.InputListValue{
+			Items: arguments,
+		}
+	} else {
+		gasLimit = nil
+		function = nil
+		args = nil
+	}
+
+	eventDataStruct := abi.StructValue{
+		Fields: []abi.Field{
+			{
+				Name:  "tx_nonce",
+				Value: abi.U64Value{Value: eventData.Nonce},
+			},
+			{
+				Name: "gas_limit",
+				Value: abi.OptionValue{
+					Value: gasLimit,
+				},
+			},
+			{
+				Name: "function",
+				Value: abi.OptionValue{
+					Value: function,
+				},
+			},
+			{
+				Name: "args",
+				Value: abi.OptionValue{
+					Value: args,
+				},
+			},
+		},
+	}
+
+	encodedOp, err := dc.serializer.Serialize([]any{eventDataStruct})
+	if err != nil {
+		return nil, err
+	}
+
+	operationBytes, _ := hex.DecodeString(encodedOp)
+	return operationBytes, nil
+}
+
+// DeserializeEventData will deserialize bytes to an event data structure
+func (dc *dataCodec) DeserializeEventData(data []byte) (*sovereign.EventData, error) {
+	if len(data) == 0 {
+		return nil, errEmptyData
+	}
+
+	nonce := &abi.U64Value{}
+	gasLimit := &abi.U64Value{}
+	function := &abi.BytesValue{}
+	args := &abi.OutputListValue{
+		ItemCreator: func() any { return &abi.BytesValue{} },
+	}
+
+	eventDataStruct := abi.StructValue{
+		Fields: []abi.Field{
+			{
+				Name:  "tx_nonce",
+				Value: nonce,
+			},
+			{
+				Name: "gas_limit",
+				Value: &abi.OptionValue{
+					Value: gasLimit,
+				},
+			},
+			{
+				Name: "function",
+				Value: &abi.OptionValue{
+					Value: function,
+				},
+			},
+			{
+				Name: "args",
+				Value: &abi.OptionValue{
+					Value: args,
+				},
+			},
+		},
+	}
+
+	err := dc.serializer.Deserialize(hex.EncodeToString(data), []any{&eventDataStruct})
+	if err != nil {
+		return nil, err
+	}
+
+	arguments := make([][]byte, 0)
+	for _, item := range args.Items {
+		arg, ok := item.(*abi.BytesValue)
+		if !ok {
+			return nil, errTypeAssertion
+		}
+
+		arguments = append(arguments, arg.Value)
+	}
+
+	var td *sovereign.TransferData
+	if len(function.Value) == 0 {
+		td = nil
+	} else {
+		td = &sovereign.TransferData{
+			GasLimit: gasLimit.Value,
+			Function: function.Value,
+			Args:     arguments,
+		}
+	}
+
+	return &sovereign.EventData{
+		Nonce:        nonce.Value,
+		TransferData: td,
+	}, nil
+}
+
+// SerializeTokenData will receive an esdt token data and serialize it
+func (dc *dataCodec) SerializeTokenData(tokenData sovereign.EsdtTokenData) ([]byte, error) {
+	tokenUris := make([]any, len(tokenData.Uris))
```

Comment:
Is it ok that you create from multiple uris ([][]byte to []byte) ?

Reply:
is array of any, `any` is `BytesValue` which is `[]byte`